File: 0_3.leetcode/java/Three_Sum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Three_Sum {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        for (int i=0;i<nums.length;i++) {
            if (i > 0 && nums[i - 1] == nums[i]) continue;
            result.addAll(find(nums, i));
        }
        return result;
    }

    private List<List<Integer>> find(int[] nums, int idx) {
        List<List<Integer>> result = new ArrayList<>();
        int target = -nums[idx];
        int left = idx + 1;
        int right = nums.length - 1;

        while (left < right) {
            int current = nums[left] + nums[right];
            if (current == target) {
                result.add(List.of(nums[idx], nums[left], nums[right]));
                left++;
                right--;

                while(left < right && nums[left] == nums[left - 1]) left++;
                while(left < right && nums[right] == nums[right + 1]) right--;
            } else if (current > target) {
                right--;
            } else {
                left++;
            }
        }
        return result;
    }

}

File: 0_3.leetcode/kotlin/src/main/kotlin/Three_Sum.kt
fun threeSum(nums: IntArray): List<List<Int>> {
    nums.sort()
    val result = mutableListOf<List<Int>>()

    for (i in nums.indices) {
        if (i > 0 && nums[i] == nums[i - 1]) continue
        result.addAll(find(nums, i))
    }

    return result
}

fun find(nums: IntArray, idx: Int): List<List<Int>> {
    val result = mutableListOf<List<Int>>()
    val target = -nums[idx]
    var left = idx + 1
    var right = nums.size - 1

    while (left < right) {
        val current = nums[left] + nums[right]
        if (current == target) {
            result.add(listOf(-target, nums[left], nums[right]))
            left++
            right--
            while (left < right && nums[left] == nums[left - 1]) left++
            while (left < right && nums[right] == nums[right + 1]) right--
        } else if (current > target) {
            right--
        } else {
            left++
        }
    }

    return result
}